[flutter_tools] Add doctor host validator

gamemaker1 opened the pull request:
## Description

Hi there! I wanted to contribute to Flutter and found this issue as a good one to start with.

Adds an extra check while running `flutter doctor` to check availability of several HTTP hosts by sending HEAD requests to them.

## Screenshot

![FlutterDoctorAddHTTPHostValidatorScreenshot](https://user-images.githubusercontent.com/34235681/111062838-791e6b00-84d1-11eb-8d28-e2430d20a62a.png)

## Issues related to this PR

Issue #49403

## Tests

Added a [test for the http host validator](https://github.com/flutter/flutter/blob/18efdb88919929c7acce96a15d0a7a5f799941f1/packages/flutter_tools/test/commands.shard/hermetic/http_host_availability_validator_test.dart).

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I am willing to follow-up on review comments in a timely manner.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

  - **I see several tests failing when I run `flutter test tests/command.shard/`, though they are not related to the HTTP Host validator.** Before making the changes, the output of running that was `11:04 +415 -44: Some tests failed.`, while afterwards, it was `10:54 +417 -44: Some tests failed.`.

- [ ] The analyzer (flutter analyze --flutter-repo) does not report any problems on my PR.

  - I had the same error appear multiple times: `import_of_legacy_library_into_null_safe`. This error only appears when I add the changes. I have currently suppressed the errors using `// ignore:` comments. Following are the packages for which the warning appears:

    In the validation class file:

    - `package:flutter_tools/src/base/io.dart`
    - `package:http/http.dart`
    - `base/platform.dart`
    - `doctor.dart`

    In the test file:

    - `package:flutter_tools/src/base/io.dart`
    - `package:flutter_tools/src/doctor.dart`
    - `package:http/http.dart`
    - `package:http/testing.dart`
    - `tests/src/common.dart`

  - I was not able to solve this problem, could you please guide me on it?

Thanks,
gamemaker1

gamemaker1 committed:
feat: add host availability check to flutter doctor

gamemaker1 committed:
chore: add test for host availability validator

gamemaker1 committed:
fix: rename HostAvailabilityValidator => HttpHostAvailablityValidator

gamemaker1 committed:
fix: rename private result class and test summaries

- renamed _HostAvailabilityResult to _HttpHostAvailabilityResult
- replaced 'host' with 'http host' in test names

gamemaker1 committed:
fix: test case for all hosts available fails

- fix check for how many non-error messages appear

gamemaker1 committed:
fix: rename validator and status strings

gamemaker1 committed:
chore: merge commits from upstream into add-doctor-host-validator

gamemaker1 committed:
chore: merge upstream commits to add-doctor-host-validator

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/http_host_availability_validator_test.dart:
```diff
@@ -0,0 +1,96 @@
+// ignore: import_of_legacy_library_into_null_safe
```
Don't ignore this error, opt the doctor validator out of null safety like the other tests

jonahwilliams commented on packages/flutter_tools/lib/src/http_host_availability_validator.dart:
```diff
@@ -0,0 +1,127 @@
+// ignore: import_of_legacy_library_into_null_safe
```
Same here, add a `// @dart = 2.9` and do not use null safety

jonahwilliams commented on packages/flutter_tools/lib/src/doctor.dart:
```diff
@@ -6,6 +6,7 @@
 
 import 'package:meta/meta.dart';
 import 'package:process/process.dart';
+import 'package:http/http.dart' as http;
```
Do not use package:http, use the io network bits instead. See base/net.dart for an example

gamemaker1 committed:
chore: merge commits from upstream into add-doctor-host-validator

gamemaker1 committed:
fix: opt out of null safety and replace http package with `base/io.dart`

gamemaker1 committed:
fix: display error message instead of calling toString on the error

gamemaker1 commented:
Hi there,

Thanks for taking the time to suggest changes. Based on your suggestions, I have pushed commits that opt out of null safety and replace the `http` package with `base/io.dart` in the validator as well as the test.

**`flutter analyze`** does not return any errors now, while **`flutter test tests/command.shard/`** returns 14 errors `04:25 +457 -14: Some tests failed` (none related to the test added by this PR).

Thanks,
gamemaker1

gamemaker1 committed:
chore: merge upstream commits into add-doctor-host-validator

gamemaker1 committed:
chore: merge upstream commits into add-doctor-host-validator

jonahwilliams commented on packages/flutter_tools/lib/src/http_host_availability_validator.dart:
```diff
@@ -0,0 +1,123 @@
+  String get slowWarning => 'HTTP host availability check is taking a long time';
+
+  /// Returns a list of URLs to check availability, different for different platforms
+  List<String> get _allRequiredHosts {
```
Technically both the pub and cloud URL can be overriden with environment variables,  see https://flutter.dev/community/china 

Only Android applications need access to maven, it might be worth excluding from this check

gamemaker1 replied:
Thanks for that, I had not thought of the environment variable overrides. I added it in 070ae59 to the validator, but did not know how to access environment variables while writing tests. Could you please guide me on that?

gamemaker1 replied:
Also, I kept the maven check, as the original issue (#49403, which references #22470 and #30597) was regarding maven not being available.

jonahwilliams replied:
Environment variables are present on the `Platform` class, you can inject this into the interface and use a FakePlatform for testing

gamemaker1 replied:
Thanks, I'll add it.

jonahwilliams commented on packages/flutter_tools/lib/src/http_host_availability_validator.dart:
```diff
@@ -0,0 +1,123 @@
+  List<String> get _allRequiredHosts {
+    if (_platform.isMacOS) {
+      return commonRequiredHostUrls + macOsRequiredHostUrls;
+    } else {
```
here and else where, if you have a return in `if/elseif`, you don't need a `else` block, just return instead like:

```
if (a) {
  return c;
}
return d;
```

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/http_host_availability_validator_test.dart:
```diff
@@ -0,0 +1,82 @@
+      }
+    });
+  });
+}
```
```suggestion
}

```

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/http_host_availability_validator_test.dart:
```diff
@@ -0,0 +1,82 @@
+        final ValidationResult result = await httpHostValidator.validate();
+
+        // Check for only one information message
+        expect(result.messages..removeWhere(
```
Here and elsewhere, i would try to make the checks a little less smart. You should start by listing out the expected messages exactly.

gamemaker1 replied:
Thanks. I changed the tests to check the type of the result [`ValidationType.installed`, `ValidationType.partial` or `ValidationType.notAvailable`].

gamemaker1 committed:
feat: check for environment variables overriding URLS

- check for PUB_HOSTED_URL and FLUTTER_STORAGE_BASE_URL environment
variables and use their values if specified

gamemaker1 committed:
fix: simplify tests to simply check the type of result

- if all hosts are available, the type will be ValidationType.installed
- if all some hosts are available and some are not, the type
will be ValidationType.partial
- if all hosts are not available, the type will
be ValidationType.notAvailable

gamemaker1 committed:
chore: merge upstream commits into add-doctor-host-validator

gamemaker1 committed:
fix: call .close on http requests made to ensure errors are caught

- catch OSErrors while making HEAD requests

gamemaker1 committed:
chore: merge upstream commits into add-doctor-host-validator

gamemaker1 committed:
fix: check environment variables from provided platform 

- (rather than from globals.dart)

gamemaker1 committed:
feat: add tests to check environment variables being detected

- env vars include PUB_HOSTED_URL and FLUTTER_STORAGE_BASE_URL

gamemaker1 committed:
chore: merge upstream commits into add-doctor-host-validator

jonahwilliams commented:
FYI:

```
The following 2 files do not have the right license header:
/b/s/w/ir/k/flutter sdk/packages/flutter_tools/lib/src/http_host_availability_validator.dart
/b/s/w/ir/k/flutter sdk/packages/flutter_tools/test/commands.shard/hermetic/http_host_availability_validator_test.dart
The expected license header is:
// Copyright 2014 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
...followed by a blank line.
```

Please make sure the tests are all passing too

gamemaker1 committed:
chore: merge upstream commits into add-doctor-host-validator

gamemaker1 committed:
fix(license): add license headers to files

gamemaker1 commented:
Thanks @jonahwilliams, I had missed out checking the test results. Based on last time's output, I:

- added the license headers to the files
- made a few changes suggested by `flutter analyze`. 

There are still two checks that fai:

- `Windows module_test` [recipe infra failure: Infra Failure: Step('Checkout flutter/flutter.git fetch') (retcode: 128)]
- `flutter-build` 

Though based on the error message, my guess is that it is not related to the changes in this PR.

Thanks,
gamemaker1

jonahwilliams commented on packages/flutter_tools/lib/src/http_host_availability_validator.dart:
```diff
@@ -0,0 +1,142 @@
+  /// Returns a list of URLs to check availability, different for different platforms
+  List<String> get _allRequiredHosts {
+    /// Hosts used by flutter on all machines
+    final List<String> commonRequiredHostUrls = <String>[
```
One more adjustment I would make - use the `Features` class to determine if android support is enabled, and do not check the host if it isn't. Same thing for iOS:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/features.dart#L43

gamemaker1 replied:
Should I check if the host platform is MacOS now that the validator is checking for the IOS feature being enabled?

gamemaker1 replied:
Done, thanks.

mouadh25 commented:
Thats consistante

gamemaker1 committed:
feat: add check for feature flags

- if android/ios is enabled as a feature, only then will the
validotor check for the respective hosts

gamemaker1 committed:
feat: add tests for feature flag check

- one each for android and ios disabled
- modify imports in validator file (make them relative)

gamemaker1 committed:
fix: do not check host platform is feature is enabled/disabled

jonahwilliams commented on packages/flutter_tools/lib/src/http_host_availability_validator.dart:
```diff
@@ -0,0 +1,141 @@
+/// Validation class that checks if all the given URLs are reachable
+class HttpHostAvailabilityValidator extends DoctorValidator {
+  HttpHostAvailabilityValidator({
+    Platform platform,
```
These fields should be annotated as `@required`

gamemaker1 replied:
Fixed, thank you!

jonahwilliams reviewed (dismissed):
LGTM with nit

gamemaker1 committed:
git: merge branch upstream commits into add-doctor-host-validator

gamemaker1 committed:
fix: add required annotations to validator

- also change import for DoctorValidator class: import it
from doctor_validator.dart

Signed-off-by: Vedant K <gamemaker0042@gmail.com>

gamemaker1 committed:
fix(test): rename import for doctor.dart to doctor_validator.dart

Signed-off-by: Vedant K <gamemaker0042@gmail.com>

gamemaker1 commented:
Hi @jonahwilliams,

It seems `Google testing` is failing, and I can't see the logs, so I don't know what cause it to fail. The `tool_tests_general` test is also failing on all three platforms, though it seems from the error messages (something about `runner_test.dart` timing out) that it is not related to this PR's changes. Please do let me know if I have to make any changes to pass the checks.

Also, thanks a lot for your guidance while I was working on this PR. Look forward to making more contributions to Flutter!

Thanks,
gamemaker1

jonahwilliams commented:
There are some failures that need to be addressed:

```
11:12 +2178 ~10 -1: test\general.shard\runner\runner_test.dart: runner error handling crash report (synchronous crash) [E]
  TimeoutException after 0:00:02.000000: Test timed out after 2 seconds.
  dart:isolate  _RawReceivePortImpl._handleMessage
  ```
  
  This looks like a serious bug - it would seem that trying to send a crash report causes a timeout in the test environment. Maybe the test is not mocking the http requests? Maybe the http requests should have some generous timeout attached?

gamemaker1 commented:
Sure, does a 10 second timeout for all HTTP requests in the test sound ok?

jonahwilliams commented:
A 10 second timeout is reasonable, but if we do timeout then instead of failing it should display a message like "The tool wasn't able to connect to X in Y seconds" with a partial success.

gamemaker1 commented:
Sorry it seems the last comment was deleted by mistake.

I wanted to ask if this is an issue with the test module or the actual validator itself, because the test uses an instance of `FakeHttpClient` to return results and doesn't make actual HTTP requests as far as I know.

jonahwilliams commented:
I'm not sure exactly, you'll need to debug the test to see where it is getting stuck

gamemaker1 committed:
git: merge upstream commits into add-doctor-host-validator

gamemaker1 committed:
feat(validator): add 10 second timeout to the network request

Signed-off-by: Vedant K <gamemaker0042@gmail.com>

gamemaker1 committed:
fix: run flutter analyze on code

Signed-off-by: Vedant K <gamemaker0042@gmail.com>

gamemaker1 committed:
git: merge upstream commits into add-doctor-host-validator

Signed-off-by: Vedant K <gamemaker0042@gmail.com>

gamemaker1 commented:
Hi @jonahwilliams,

I tried hard to reproduce the error but was unsuccesful. I tried running the entire general shard and the individual test (`test/general.shard/runner/runner_test.shard`) locally, but did not succeed in getting that error. I looked up the error (`dart:isolate  _RawReceivePortImpl._handleMessage`), and it seems that is the last line of the stack trace of any async error, not particularly related to network requests.

My initial changes to the Doctor HTTP host validator lengthening the timeout for the network requests had no effect either, let me know if I should roll that back.

Also, the checks on Github seem to be queued for a few hours now, not sure why.

Thanks,
gamemaker1

jonahwilliams commented:
I'm not sure why its failing TBH, I think you'll need to debug the tests to see what expectations are being missed

gamemaker1 committed:
git: merge upstream commits into add-doctor-host-validator

Signed-off-by: Vedant K <gamemaker0042@gmail.com>

jonahwilliams commented on packages/flutter_tools/lib/src/http_host_availability_validator.dart:
```diff
@@ -0,0 +1,148 @@
+    try {
+      // Make the HEAD request
+      final HttpClientRequest headReq = await _httpClient.headUrl(Uri.parse(hostUrl));
+      await headReq.close().timeout(const Duration(seconds: 10));
```
Try making the timeout configurable and then set it to a fairly low time for testing?

jonahwilliams commented:
Because of the lack of activity on this PR I'm going to close it to remove from our review queue. Please feel free to send us a new PR or discuss in an issue anytime.
